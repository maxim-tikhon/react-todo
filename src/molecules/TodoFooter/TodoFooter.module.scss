.todoFooter {
	display: flex;
	justify-content: space-between;
	width: 100%;
 
  .totalCount {
    color: grey;
    font-size: 14px;
    width: 125px;
  }
 
  .buttonsContainer {
    > * {
      margin-right: 10px;
    }
  }
}

@media (max-width: 768px) {
	.todoFooter {
		flex-direction: column;
    align-items: flex-start;
	}
}
