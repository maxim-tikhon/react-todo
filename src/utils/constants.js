export const FILTER = {
	ALL: 'all',
	ACTIVE: 'active',
	COMPLETED: 'completed',
};

export const TODO_STOARGE_NAME = 'tasks';
