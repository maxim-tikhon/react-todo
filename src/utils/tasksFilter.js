const taskFilter = {
	filterTasks(tasks, filter) {
		switch (filter) {
			case "all":
				return tasks;
			case "active":
				return tasks?.filter((task) => !task.completed);
			case "completed":
				return tasks?.filter((task) => task.completed);
			default:
				return tasks;
		}
	}
};

export default taskFilter;
