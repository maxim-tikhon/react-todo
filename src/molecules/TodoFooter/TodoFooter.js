import LinkButton from "../../atoms/LinkButton/LinkButton";
import { FILTER } from "../../utils/constants";
import styles from "./TodoFooter.module.scss";

const TodoFooter = ({ numberOfActiveTasks, filter, onChangeFilter, onClearCompletedTasks }) => {
	return (
		<div className={styles.todoFooter}>
			{numberOfActiveTasks > 0 ? (
				<span className={styles.totalCount}>
					{numberOfActiveTasks} {numberOfActiveTasks === 1 ? "task" : "tasks"} left
				</span>
			) : (
				<span className={styles.totalCount}>All tasks completed</span>
			)}
			<div className={styles.buttonsContainer}>
				<LinkButton isActive={filter === FILTER.ALL} onClick={() => onChangeFilter(FILTER.ALL)}>
					All
				</LinkButton>
				<LinkButton isActive={filter === FILTER.ACTIVE} onClick={() => onChangeFilter(FILTER.ACTIVE)}>
					Active
				</LinkButton>
				<LinkButton isActive={filter === FILTER.COMPLETED} onClick={() => onChangeFilter(FILTER.COMPLETED)}>
					Completed
				</LinkButton>
			</div>
			<LinkButton onClick={onClearCompletedTasks}>Clear Completed</LinkButton>
		</div>
	);
};

export default TodoFooter;
