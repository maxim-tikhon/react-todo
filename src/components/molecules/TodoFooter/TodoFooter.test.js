import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import TodoFooter from './TodoFooter';
import { FILTER } from '../../../utils/constants';

describe('TodoFooter', () => {
  const mockNumberOfActiveTasks = 2;
  const mockFilter = FILTER.ACTIVE;
  const mockOnChangeFilter = jest.fn();
  const mockOnClearCompletedTasks = jest.fn();

  it('renders correctly', () => {
    const output = renderer.create(<TodoFooter numberOfActiveTasks={3} filter={FILTER.ACTIVE} />).toJSON();
    expect(output).toMatchSnapshot();
  });

  test('displays correct text when numberOfActiveTasks = 0', () => {
    render(<TodoFooter numberOfActiveTasks={0} />);

    const totalCount = screen.getByText('All tasks completed');

    expect(totalCount).toBeInTheDocument();
  });

  test('displays correct text when numberOfActiveTasks = 1', () => {
    render(<TodoFooter numberOfActiveTasks={1} />);

    const totalCount = screen.getByText('1 task left');

    expect(totalCount).toBeInTheDocument();
  });

  test('displays correct text when numberOfActiveTasks > 1', () => {
    render(<TodoFooter numberOfActiveTasks={mockNumberOfActiveTasks} />);

    const totalCount = screen.getByText('2 tasks left');

    expect(totalCount).toBeInTheDocument();
  });

  test('displays 3 filter links (All, Active, Completed)', () => {
    render(<TodoFooter />);

    const allFilterLink = screen.getByText('All');
    const activeFilterLink = screen.getByText('Active');
    const completedFilterLink = screen.getByText('Completed');

    expect(allFilterLink).toBeInTheDocument();
    expect(activeFilterLink).toBeInTheDocument();
    expect(completedFilterLink).toBeInTheDocument();
  });

  test('calls onChangeFilter callback with correct value when filter link is clicked', () => {
    render(<TodoFooter filter={mockFilter} onChangeFilter={mockOnChangeFilter} />);

    const allFilterLink = screen.getByText('All');
    const activeFilterLink = screen.getByText('Active');
    const completedFilterLink = screen.getByText('Completed');

    fireEvent.click(allFilterLink);
    fireEvent.click(activeFilterLink);
    fireEvent.click(completedFilterLink);

    expect(mockOnChangeFilter).toHaveBeenCalledTimes(3);
    expect(mockOnChangeFilter).toHaveBeenNthCalledWith(1, FILTER.ALL);
    expect(mockOnChangeFilter).toHaveBeenNthCalledWith(2, FILTER.ACTIVE);
    expect(mockOnChangeFilter).toHaveBeenNthCalledWith(3, FILTER.COMPLETED);
  });

  test('calls onClearCompletedTasks callback when "Clear Completed" link is clicked', () => {
    render(<TodoFooter onClearCompletedTasks={mockOnClearCompletedTasks} />);

    const clearCompletedLink = screen.getByText('Clear Completed');

    fireEvent.click(clearCompletedLink);

    expect(mockOnClearCompletedTasks).toHaveBeenCalledTimes(1);
  });
});
