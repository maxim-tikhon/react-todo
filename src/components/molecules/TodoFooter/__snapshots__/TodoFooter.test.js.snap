// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoFooter renders correctly 1`] = `
<div
  className="todoFooter"
>
  <span
    className="totalCount"
  >
    3
     
    tasks
     left
  </span>
  <div
    className="buttonsContainer"
  >
    <button
      className="linkButton"
      onClick={[Function]}
      type="button"
    >
      All
    </button>
    <button
      className="linkButton active"
      onClick={[Function]}
      type="button"
    >
      Active
    </button>
    <button
      className="linkButton"
      onClick={[Function]}
      type="button"
    >
      Completed
    </button>
  </div>
  <button
    className="linkButton"
    type="button"
  >
    Clear Completed
  </button>
</div>
`;
